nep-17.mediawiki

```diff
@@ -0,0 +1,107 @@
+<pre>
+  NEP: 17
+  Title: Token Standard
+  Author: Erik Zhang <erik@neo.org>
+  Type: Standard
+  Status: Draft
+  Created: 2020-11-12
+  Replaces: 5
+</pre>
+
+==Abstract==
+
+This proposal outlines a token standard for the NEO blockchain that will provide systems with a generalized interaction mechanism for tokenized Smart Contracts.  This mechanic, along with the justification for each feature are defined.  A template and examples are also provided to enable the development community.
+
+==Motivation==
+
+As the NEO blockchain scales, Smart Contract deployment and invocation will become increasingly important.  Without a standard interaction method, systems will be required to maintain a unique API for each contract, regardless of their similarity to other contracts.  Tokenized contracts present themselves as a prime example of this need because their basic operating mechanism is the same.  A standard method for interacting with these tokens relieves the entire ecosystem from maintaining a definition for basic operations that are required by every Smart Contract that employs a token.
+
+==Specification==
+
+In the method definitions below, we provide both the definitions of the functions as they are defined in the contract as well as the invoke parameters.
+
+===Methods===
+
+====symbol====
+
+<pre>
+public static string symbol()
+</pre>
+
+Returns a short string symbol of the token managed in this contract. e.g. <code>"MYT"</code>. This symbol SHOULD be short (3-8 characters is recommended), with no whitespace characters or new-lines and SHOULD be limited to the uppercase latin alphabet (i.e. the 26 letters used in English).
+
+This method MUST always return the same value every time it is invoked.
+
+====decimals====
+
+<pre>
+public static byte decimals()
+</pre>
+
+Returns the number of decimals used by the token - e.g. <code>8</code>, means to divide the token amount by <code>100,000,000</code> to get its user representation.
```

Comment:
The NeoVM limit integers to 256-bits.

Reply:
True, but people are interacting with the node via JSON RPC and while it technically doesn't limit numbers in any way, in practice there is `MAX_SAFE_INTEGER` in the C# code. I'd say it's too restrictive (although it follows RFC 7159 recommendation) because it doesn't allow even for 100M with 8 decimals (which is easily reached for GAS), but at the same time even if we're to remove this limitation (it's not hard to do that), probably we'll have client compatibility problems with anything not fitting in at least int64 limitations.

Reply:
I tend to agree with @roman-khimov, but I've seen many people using 16 already... so I agree we may put some limit, but the point is: based on what?
int64 is a lot of numbers, but it's not clear to me if a hyperinflationary coin could beat that... is there any risk for that?

Reply:
Maybe we should optimize RpcServer to recognize the number represented by the string. But there is no need to modify the standard.

Reply:
The main problem is that it's not just about RpcServer (as I've said, it's not hard to fix that), _every other client_ (or server) needs to support that too and I'm not sure we can expect that to happen.

Reply:
Is there any suggestion? I don't think it's easy to limit it here.

Reply:
I doubt anyone goes beyond 8 here, but 12 seems to be viable (picocoins) and that's what we're using for internal NeoFS calculations now (after failing with 16). As a reference, Neo 2 UTXO-based tokens were limited to 8 here with int64 values.

So, 12 here and int64 limits for balance/totalsupply, seems fine to me. It also allows for some optimizations in native tokens replacing big integer calculations with int64.

Reply:
@roman-khimov What I've seen regarding precision, I guess guys from DEX (maybe Nash) require that, is that 16 precision is mostly necessary in many scenarios (the reason is connected to the same principle of "gas distribution" that also made very difficult to transform NEO into fractional units). So, 16 precision is indeed necessary, but I've never seen more than that. Regarding the total supply, I do not imagine an upper limit on that, but maybe there's a "reasonable number".. I don't know.

Reply:
After some additional research I'd agree that 16 really is in use on other chains and we shouldn't place any limits in this standard. This still leaves us with JSONization problem, though.

Reply:
https://github.com/neo-project/neo/issues/2082